src/kw_ssh.sh

```diff
@@ -109,6 +122,146 @@ function run_command_in_the_remote()
   cmd_manager "$flag" "$ssh_compose"
 }
 
+# This function transfers a local file or directory to a
+# specified remote destination.
+function ssh_send_file_to_remote()
+{
+  local user
+  local port
+  local remote
+  local remote_file
+  local remote_file_host
+  local send_path
+  local to_path
+  local flag
+  local cmd
+  local send_option
+  local ssh_compose
+  local filename
+  local filename_path
+
+  flag=${flag:-'SILENT'}
+
+  user="${remote_parameters['REMOTE_USER']}"
+  port="${remote_parameters['REMOTE_PORT']}"
+  remote="${remote_parameters['REMOTE_IP']}"
+  remote_file="${remote_parameters['REMOTE_FILE']}"
+  remote_file_host="${remote_parameters['REMOTE_FILE_HOST']}"
+  send_path=${options_values['SEND']}
+  to_path=${options_values['TO']:-'.'}
+  flag=${options_values['VERBOSE']}
+  ssh_compose=$(handle_ssh)
+  filename=$(basename "$send_path")
+  filename_path=$(printf "%s/%s" "$to_path" "$filename" | tr --squeeze-repeats '/')
+
+  cmd_manager "$flag" "${ssh_compose} [[ -f '${to_path}' ]]"
+
+  if [[ "$?" -eq 0 ]]; then
+    complain "The '${to_path}' file already exists on the remote machine."
+    if [[ $(ask_yN 'Do you wish to overwrite it?') =~ '0' ]]; then
+      exit 22 # EINVAL
+    fi
+  fi
+
+  cmd_manager "$flag" "${ssh_compose} [[ -f '${to_path}/${filename}' ]]"
+
+  if [[ "$?" -eq 0 ]]; then
+    complain "The '${filename_path}' file already exists on the remote machine."
+    if [[ $(ask_yN 'Do you wish to overwrite it?') =~ '0' ]]; then
+      exit 22 # EINVAL
+    fi
+  fi
+
+  if [[ ! -e "$send_path" ]]; then
+    complain "The file or directory '${send_path}' does not exist."
+    exit 22 # EINVAL
+  fi
+
+  if [[ -f "$remote_file" ]]; then
+    cmd="rsync -e 'ssh -F ${remote_file}' -avz ${send_path} ${remote_file_host}:${to_path} > /dev/null 2>&1"
+  else
+    cmd="rsync -e 'ssh -p ${port}' -avz ${send_path} ${user}@${remote}:${to_path} > /dev/null 2>&1"
+  fi
+
+  cmd_manager "$flag" "$cmd"
+
+  if [[ "$?" -ne 0 ]]; then
+    complain "An error occurred while uploading the file(s)."
+    exit 22 # EINVAL
```

Comment:
Check the `Exit Values` section on the rsync page; you will see multiple return errors. It will be better if you return the rsync code, and, if possible, print a detailed error message based on the rsync man page.

Reply:
I can do that, but wouldn't it take up a lot of space in the code? I think it would look more or less like this:

```
 rsync_exit_code=$?
  if [[ "$rsync_exit_code" -ne 0 ]]; then
    case $rsync_exit_code in
      0)
        error_message="Success"
        ;;
      1)
        error_message="Syntax or usage error"
        ;;
      2)
        error_message="Protocol incompatibility"
        ;;
      3)
        error_message="Errors selecting input/output files or directories"
        ;;
      4)
        error_message="Requested action not supported: manipulating 64-bit files on incompatible platform or using unsupported option"
        ;;
      5)
        error_message="Error starting client-server protocol"
        ;;
      6)
        error_message="Daemon unable to append to log-file"
        ;;
      10)
        error_message="Error in socket I/O"
        ;;
      11)
        error_message="Error in file I/O"
        ;;
      12)
        error_message="Error in rsync protocol data stream"
        ;;
      13)
        error_message="Errors with program diagnostics"
        ;;
      14)
        error_message="Error in IPC code"
        ;;
      20)
        error_message="Received SIGUSR1 or SIGINT"
        ;;
      21)
        error_message="Some error returned by waitpid()"
        ;;
      22)
        error_message="Error allocating core memory buffers"
        ;;
      23)
        error_message="Partial transfer due to error"
        ;;
      24)
        error_message="Partial transfer due to vanished source files"
        ;;
      25)
        error_message="The --max-delete limit stopped deletions"
        ;;
      30)
        error_message="Timeout in data send/receive"
        ;;
      35)
        error_message="Timeout waiting for daemon connection"
        ;;
      *)
        error_message="Unknown error"
        ;;
    esac

    complain "An error occurred while uploading the file(s). Error code: $rsync_exit_code - $error_message"
    exit $rsync_exit_code
  fi
```